### ISSUE TITLE
4.14.1 Release checklist

### ISSUE THREAD

**Issue by compulim** [microsoft/BotFramework-WebChat#4044]
# Checklist

## Build

1. [x] ~Bump MockBot to Bot Framework SDK release 4.14.1~
   - https://www.npmjs.com/package/botbuilder/v/4.14.1
1. [x] ~Bump `botframework-directlinejs` to `0.14.1`~
   - No newer version
1. [x] Bump to `4.14.1`
   - [x] Update `CHANGELOG.md` to mark specific changes in `4.14.1`
   - [x] Run `npm version --no-git-tag-version 4.14.1`
   - [x] Merged into `master`, the PR number is #4045
   - Commit is `1b842df`
   - Do not merge any other unrelated changes after this PR. Any other PR merged, will need to be re-tested
1. [x] Run daily pipeline manually, set "generate release version number" to `true`
   - (This will not push to NPM or CDN)
   - Pipeline name is `BotFramework-WebChat-daily`
   - The build number is [`267509`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=267509&view=results) and commit is `1b842df`
1. [x] Wait for [`WebChat-release-testing` pipeline](https://fuselabs.visualstudio.com/BotFramework-WebChat/_release?_a=releases&view=mine&definitionId=17) to complete
   - Pipeline name is `Push-Release-Testing-to-GitHub-Pages`
   - The release ID is [`2777`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=2777)
1. [x] ~Add manual tests to WebChat-release-testing as needed~

## Test

> The test should run against the build artifacts from Azure Pipelines.

1. [x] Manual testing on major browsers using `webchat-release-testing`
   - [x] Before starting testing, update all the browser version to latest
   - [x] Chrome 93.0.4577.63
   - [x] Edge 92.0.902.84
   - [x] Firefox 91.0.2
   - [x] IE11 (IE Mode under Windows 11 22000.168)
   - [x] macOS Safari 14.1.1 (16611.2.7.1.4)
   - [x] iOS Safari 14.7.1
   - [x] Android Chrome 92.0.4515.159
1. [x] ~Test specific fixes related to `4.14.1` and previous releases~
   
   - Test Markdown on IE11

## Release

1. [x] Verify on WebChat-release-testing
1. [x] Make sure you are on `main` ~or `qfe`~ branch, run `git status` to check
1. [x] `git pull`
1. [x] Verify `/package.json`, `/package-lock.json`, and `CHANGELOG.md` has a version of `4.14.1`
1. [x] `git log`
   - Verify the latest commit is `1b842df`
1. [x] `git tag v4.14.1`
1. [x] `git push -u upstream v4.14.1`
   - You do not need to kick off a build again, use the previous build
1. [x] Create a new GitHub release, copy entries from `CHANGELOG.md`
   - [x] Subresource Integrity can be generated by
      - From local: `cat webchat.js | openssl dgst -sha384 -binary | openssl base64 -A`
      - From CDN: `curl -H 'Accept-Encoding: gzip' https://cdn.botframework.com/botframework-webchat/4.14.1/webchat.js | gunzip - | openssl dgst -sha384 -binary | openssl base64 -A`
   - [x] Attach assets including 3 JS files, `stats.json` and 5 tarballs
      - You can copy the artifacts from `webchat-release-testing/drops`
      - Tarballs download from npmjs
         ```
         curl -LO https://registry.npmjs.org/botframework-directlinespeech-sdk/-/botframework-directlinespeech-sdk-4.14.1.tgz
         curl -LO https://registry.npmjs.org/botframework-webchat/-/botframework-webchat-4.14.1.tgz
         curl -LO https://registry.npmjs.org/botframework-webchat-core/-/botframework-webchat-core-4.14.1.tgz
         curl -LO https://registry.npmjs.org/botframework-webchat-api/-/botframework-webchat-api-4.14.1.tgz
         curl -LO https://registry.npmjs.org/botframework-webchat-component/-/botframework-webchat-component-4.14.1.tgz
         ```
1. [x] Kick off release to NPM
   - Release name is `[[PROD]]Push-WebChat-to-npmjs`
   - The build number is [267509](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=267509&view=results) release number is [`35`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=2778) and commit is `1b842df`
   - [x] Verify package content then click Resume
   - [x] Retain the release indefinitely
1. [x] Kick off release to CDN (cutoff at 2PM PST, Mon-Thu only)
   1. [x] Prepare the email for approval
   - Release name is `[[PROD]]Push-WebChat-to-Prod-CDN-with-approval`
   - The build number is [`267509`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=267509&view=results), release number is [`40`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=2779) and commit is `1b842df`
   - Script build number is `131156` (this is fixed)
   1. [x] Send reminder email to approvers
   - Retain the build indefinitely

## Post-release verification - complete within 30 minutes after release to NPM

- [x] Test using `webchat-release-testing`
   1. [x] Clone https://github.com/corinagum/WebChat-release-testing/
   1. [x] `01.create-react-app`
      1. [x] Nuke `01.create-react-app/node_modules`
      1. [x] `npm install`
      1. [x] `npm install botframework-webchat@4.14.1` (just install the bundle package)
      1. [x] `npm run build`
   1. [x] Others
      -  Using script tags from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.14.1, with subresource integrity
         ```
         <script
           crossorigin="anonymous"
           integrity="sha384-uJWBe8iPb6wN8/AaLQ0UKjgYEtWY9uIVgpkAID5kUNo2sV9hYmFbTDQXeRXyq7bl"
           src="../drops/webchat.js"
         ></script>

         <script
           crossorigin="anonymous"
           integrity="sha384-R4Eevwf75H8RQkTqCaFNpvO0V9nq+heLX30jLGKNrAICynjuHUbMo7D2usohGsRu"
           src="../drops/webchat-es5.js"
         ></script>

         <script
           crossorigin="anonymous"
           integrity="sha384-Xyp6O3TOCrP+PVhd15I1Ehs8QTga7zqc/KYj6Idr9MbR10/SX5p6pSBhTAvAT/Sd"
           src="../drops/webchat-minimal.js"
         ></script>
         ```
   1. [x] `npx serve` (at repo root)
   1. [x] Go to http://localhost:5000/ to test, including IE11

## Notification to interested parties

- [x] Update partner page on Adaptive Cards doc
   - https://docs.microsoft.com/en-us/adaptive-cards/resources/partners
   - PR at https://github.com/MicrosoftDocs/AdaptiveCards/pull/413
- [x] Notify related parties for the following fixes
   - [x] SDK team
   - [x] Omnichannel
   - [x] Pooja/Zhipeng
- [x] ~Update root README.md with feature notes -- Note: PR will be combined with post-release checklist PR~ (No new features on 4.14.1 patch release)
- [x] ~Update https://github.com/microsoft/BotFramework-sdk/issues/XXX~
   
---

# Post-release checklist

These are chores that we should do before starting the cycle to reduce ripple effects if we do it in mid-cycle.

Tips: 

- Clean your repo before start
- Remove `node_modules` from all folder
   - `git clean -fdx`
- Never delete `package-lock.json`
- If you mess it up, tableflip and redo
- In `component/package.json`
   - Remove reference to `botframework-webchat-core` by hand-modifying `package.json`
   - Then, `npm install` (symlinks will be broken afterward)
   - Then, add those references back by hand-modifying `package.json`
   - This also applies for other packages with similar dependencies/symlinks
   - To build afterward, do tableflip to rebuild those symlinks

## Applies to all releases

> This list should be copied to versions in the future.

- [x] ~If on QFE branch, make sure `CHANGELOG.md` and version number bump is cherry-picked to `master`~
   - [x] ~`git checkout master`~
   - [x] ~`git cherry-pick XXX` (the commitish for bumping version number and `CHANGELOG.md`)~
- [x] ~Correct the date for 4.14.1 in `CHANGELOG.md` in PR #XXX~
   - There could be last minute fixes that could push the planned date later than the one in `CHANGELOG.md`
- [x] Bump `package.json` to `4.14.2-0` in PR #4048
   - Run `npm version prepatch --no-git-tag-version`
- [x] Update [`servicingPlan.json`](https://github.com/microsoft/BotFramework-WebChat/blob/master/packages/embed/servicingPlan.json) in PR #4048
   - Add deprecation notes for previous versions
   - Subresource integrity hash from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.14.1
- [x] Update all samples to use `4.14.1` in PR #4048
   - Some samples are pointing to GitHub Releases because the sample need new features from daily build
   - Search "https://github.com/microsoft/BotFramework-WebChat/releases/download/"
      - And replace with "https://cdn.botframework.com/botframework-webchat/latest/"
- [x] Clean up unnecessary branch on official repo
- [x] Understand production-hitting vulnerabilities
   - [x] Create a new folder
   - [x] Run `npm init` with default values
   - [x] Run `npm install botframework-webchat@4.14.1`
   - [x] Look at the result and see if there are any production-hitting vulnerabilities, investigate if needed

The PR is https://github.com/microsoft/BotFramework-WebChat/pull/XXX

### Bump all dependencies to latest version

On each `package.json` under root and `/packages/`:

- [x] ~All NPM packages (`@babel/*`, `jest`, `lerna`, `typescript`, `webpack`, etc) in PR #XXX~
   - Don't bump the following packages
      - `react`, `react-dom`, `redux`, `redux-saga`, `@types/react`
      - `rxjs@5` on `playground`
      - `microsoft-cognitiveservices-speech-sdk`
        - We will bump it in a separate PR
      - `adaptivecards`
        - We will bump it in a separate PR
      - `expect@25` under `test/harness`
        - The newer version requires Node.js (`graceful-fs`)
- [x] ~Run `npm audit fix` to make sure everything is fixed~
- [x] ~Test under IE11 to make sure all dependencies are working~
- [x] ~List steps to verify bumping `microsoft-cognitiveservices-speech-sdk`~

#### Obstacles to bump `npm`

As of writing, `npm@7.19.1` (bundled by `node@16`) has issue on running `postinstall` script.

Also, `microsoft-cognitiveservices-speech-sdk@1.17.0` is peer-depends on `@angular/common`, which `npm@7` will automatically install missing peer dependencies. [Track the issue here](https://github.com/microsoft/cognitive-services-speech-sdk-js/issues/403).

We need at least both issues to be fixed before we can jump to `npm@7`.

### Bump Docker image

> The Docker image can be found at root `docker-compose.yml` and `Dockerfile*`.

- [x] ~Docker container for headless Chrome (#XXX)~
   - They recently moved from `3.14.159-xxx` tag scheme to a more sensible `87.0` tag scheme
   - Tags can be found at https://hub.docker.com/r/selenium/node-chrome/tags
   - Preferably in separate PR because screenshots change can be large occasionally
   - Run tests locally, as the screenshots can be slightly different
